### PR TITLE
[Tracing] Fix issues when compiling & instrumenting twice

### PR DESCRIPTION
**Pull request by nickgg**
*Description*: When auto instrumenting we add Placeholders to the Module to hold the backing tensors for the instrumented metrics. When compiling twice we end up creating multiple placeholders, some which that don't have an associated weight in the IR which breaks when collecting constants. This fixes auto instrumentation by reusing a previous backing tensor if the size matches.
*Testing*: new unit tests, asan.
*Documentation*:

### CODE REVIEW

**Comment by nickgg** [comment]
Remove std::chrono_literals as it is not supported by our cave-paintings-based compilation target.

File: lib/Backends/Backend.cpp [hunk 1]
```diff
@@ -71,27 +71,65 @@ void Backend::autoInstrument(TraceInfo &traceInfo, IRFunction *IR) const {
...
+  auto &varmap = IR->getVariableMap();
+  auto type = F->getParent()->uniqueType(
       ElemKind::Int64ITy,
-      {instructions.size() + 1,
-       getTraceEventDataSize() / Type::getElementSize(ElemKind::Int64ITy)},
-      F->getName().str() + "_instrumentation", false);
+      {numEvents,
+       getTraceEventDataSize() / Type::getElementSize(ElemKind::Int64ITy)});
+
+  WeightVar *backingWeight = nullptr;
```
Comment (jfix71) [inline lib/Backends/Backend.cpp]: Can you add some comments here about what's going on for the different combinations here (`backingWeight` true/false, and `backingPH` true/false).

File: lib/Backends/Backend.cpp [hunk 2]
```diff
@@ -71,27 +71,65 @@ void Backend::autoInstrument(TraceInfo &traceInfo, IRFunction *IR) const {
...
+
+  if (backingPH) {
+    auto it = varmap.find(backingPH);
+    if (it != varmap.end() && backingPH->getType()->isEqual(type)) {
+      backingWeight = llvm::cast<WeightVar>(it->second);
+    } else {
+      // This isn't ideal, the placeholder exists but we have no weight.
+      // Probably indicates a bug in the graph, best we can do is create a new
+      // placeholder and weight for the instrumentation.
+      backingPH = nullptr;
```
[the remark this answers is not in the document]
Comment (nickgg) [reply lib/Backends/Backend.cpp]: Nope, I could add an assert here

**Comment by jfix71** [comment]
> I had only tested this on the CPU and Interpreter backends which run in order, but theres no reason it has to work like that. Maybe we should have the TraceEventNode take in an input so it is attached to the end time of a specific instruction.

Well this is independent of the backend -- the instructions could be scheduled in any order that respects the dependences. E.g. in `Interpreter/TraceEventsTest.internalGap/0`, I took a look at the schedule from the test (via cmd line option `-dump-ir`) -- you can see the two `traceevents` are immediately scheduled:

```
code {
  0 %interpret_middle_section_instrumentation = traceevent @in %A { Index: 0}
  1 %interpret_middle_section_instrumentation1 = traceevent @in %A { Index: 1}
  2 %conv1_res = allocactivation  { Ty: float<1 x 32 x 32 x 16>} // size: 65536 // Users: @in 5, @out 6, @out 3
  3 %conv1 = convolution @out %conv1_res, @in %input, @in %filter, @in %bias { Kernels: [5, 5], 
  ...
}
```

So yeah I think giving them an input would make sense here, so they won't be scheduled until the `NodeValue` they're expecting is complete. We may even need to adjust the scheduler to give priority to these Instructions, because otherwise they could be left to all come last in the schedule too...?

**Comment by nickgg** [comment]
@jfix71 made a new issue #2535 for that fix.

File: tests/unittests/TraceEventsTest.cpp [hunk 3]
```diff
@@ -313,6 +318,64 @@ TEST_P(TraceEventsTest, manualAndAutomatic) {
   ASSERT_EQ(manualEvents, numEvents);
 }
 
+TEST_P(TraceEventsTest, twoCompiles) {
```
[the remark this answers is not in the document]
Comment (nickgg) [reply tests/unittests/TraceEventsTest.cpp]: actually, most of our tests don't have doxygen comments - this doesn't seem like a standard across the project (see: any other test in that file, ProvisionerTest, BackendCorrectnessTest, most of OperatorTest, etc).

Is doxy comments for test cases particularly useful? Why would someone be looking at the doxygen output of the unittests?

[the remark this answers is not in the document]
Comment (jfix71) [reply tests/unittests/TraceEventsTest.cpp]: Hmm -- looks like we have comments on about a third of the tests. Definitely inconsistent here.

I don't think they're necessarily useful for actually processing with doxygen. I mostly just like them so I can quickly understand what the intention of the test is at a high level.

[the remark this answers is not in the document]
Comment (nickgg) [reply tests/unittests/TraceEventsTest.cpp]: For this PR how about I add one for this test and we can address the remaining later?

**Pull request merged**